runtime/src/snapshot_utils.rs

```diff
@@ -471,6 +472,102 @@ pub fn move_and_async_delete_path(path: impl AsRef<Path> + Copy) {
+        fs::read_dir(&account_hardlinks_dir)
+            .unwrap_or_else(|_| {
+                panic!(
+                    "Unable to read snapshot hardlink directory: {}",
+                    account_hardlinks_dir.display()
+                )
+            })
+            .for_each(|entry| {
+                let entry = entry.unwrap_or_else(|_| {
+                    panic!(
+                        "Unable to read snapshot hardlink directory entry: {}",
+                        account_hardlinks_dir.display()
+                    )
+                });
+                let path = entry.path();
+                let target = fs::read_link(&path).unwrap_or_else(|_| {
+                    panic!(
+                        "Unable to read snapshot hardlink directory entry: {}",
+                        path.display()
+                    )
+                });
+                account_snapshot_dirs_referenced.insert(target);
+            });
+    }
+
+    // loop through the account snapshot hardlink directories, if the directory is not in the account_snapshot_dirs_referenced set, delete it
+    for account_snapshot_path in account_snapshot_paths {
+        fs::read_dir(account_snapshot_path)
+            .unwrap_or_else(|_| {
+                panic!(
+                    "Unable to read snapshot hardlink directory: {}",
+                    account_snapshot_path.display()
+                )
+            })
+            .for_each(|entry| {
+                let entry = entry.unwrap_or_else(|_| {
+                    panic!(
+                        "Unable to read snapshot hardlink directory entry: {}",
+                        account_snapshot_path.display()
+                    )
+                });
+                let path = entry.path();
+                if !account_snapshot_dirs_referenced.contains(&path) {
+                    info!(
+                        "Removing unreferenced account snapshot hardlink directory: {}",
+                        path.display()
+                    );
+                    move_and_async_delete_path(&path);
+                }
+            });
+    }
+}
+
+// For all account_paths, set up the run/ and snapshot/ sub directories.
+// If the sub directories do not exist, the account_path will be cleaned because older version put account files there
+pub fn set_up_account_run_and_snapshot_paths(
+    account_paths: &[PathBuf],
+) -> (Vec<PathBuf>, Vec<PathBuf>) {
+    // create the run/ and snapshot/ sub directories for each account_path
+
+    let account_run_and_snapshot_paths: Vec<(PathBuf, PathBuf)> = account_paths
+        .iter()
+        .map(|account_path| {
+            match fs::create_dir_all(account_path).and_then(|_| fs::canonicalize(account_path)) {
+                Ok(account_path) => account_path,
+                Err(err) => {
+                    eprintln!("Unable to access account path: {account_path:?}, err: {err:?}");
```

Comment:
@apfitzge A CI test failed.  It looks that eprintln is not allowed inside runtime/src/*.

Reply:
Option 1:
let the function return SnapshotError::InvalidAccountPath.  But it is in the closure in the .map().map() chain.  I need to change the closure return type to Result, and unwrap it in the .map chain.  it gets complicated.
Option 2:
move the create_dir_all back to ledger main and validator main.  a little extra duplication.
Option 3:
use panic!, not eprintln!.

Not sure yet which option is the best.

Reply:
changed to panic! for now to pass the CI check, so I can test the run the rest of the CI tests on the PR.

Reply:
imo we should definitely choose option 2.

The complexity comes from having multiple maps, which we can just not do. Just use a single map and ? away errors within that closure.